tsdb/engine/wal/wal.go

```diff
@@ -0,0 +1,1013 @@
+	// will affect what percentage of the WAL gets compacted at a time. For instance, a setting of 10 means
+	// we generally will be compacting about 10% of the WAL at a time.
+	PartitionCount uint64
+
+	IndexWriter indexWriter
+}
+
+// indexWriter is an interface for the indexed database the WAL flushes data to
+type indexWriter interface {
+	// time ascending points where each byte array is:
+	//   int64 time
+	//   data
+	WriteIndex(pointsByKey map[string][][]byte) error
+}
+
+func NewLog(path string) *Log {
+	return &Log{
+		path:  path,
+		flush: make(chan int, 1),
+
+		// these options should be overriden by any options in the config
+		LogOutput:              os.Stderr,
+		FlushInterval:          DefaultFlushInterval,
+		SegmentSize:            DefaultSegmentSize,
+		MaxSeriesSize:          DefaultMaxSeriesSize,
+		CompactionThreshold:    DefaultCompactionThreshold,
+		PartitionSizeThreshold: DefaultPartitionSizeThreshold,
+		ReadySeriesSize:        DefaultReadySeriesSize,
+		PartitionCount:         DefaultPartitionCount,
+	}
+}
+
+// Open opens and initializes the Log. Will recover from previous unclosed shutdowns
+func (w *Log) Open() error {
+	// open the partitions
+	w.partitions = make(map[uint8]*partition)
+	for i := uint64(1); i <= w.PartitionCount; i++ {
+		p, err := newPartition(uint8(i), w.path, w.SegmentSize)
+		if err != nil {
+			return err
+		}
+		w.partitions[uint8(i)] = p
+	}
+	if err := w.openPartitionFiles(); err != nil {
+		return err
+	}
+
+	w.logger = log.New(w.LogOutput, "[wal] ", log.LstdFlags)
+
+	w.flushCheckTimer = time.NewTimer(flushCheckInterval)
+
+	// Start background goroutines.
+	w.wg = *&sync.WaitGroup{}
+	w.wg.Add(1)
+	w.closing = make(chan struct{})
+	go w.autoflusher(w.closing)
+
+	return nil
+}
+
+// Cursor will return a cursor object to Seek and iterate with Next for the WAL cache for the given
+func (w *Log) Cursor(key string) tsdb.Cursor {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+
+	p := w.walPartition([]byte(key))
+
+	return p.cursor(key)
+}
+
+// partition is a set of files for a partition of the WAL. We use multiple partitions so when compactions occur
+// only a portion of the WAL must be flushed and compacted
+type partition struct {
+	id                 uint8
+	path               string
+	mu                 sync.Mutex
+	currentSegmentFile *os.File
+	currentSegmentSize int
+	currentSegmentID   uint32
+	lastFileID         uint32
+	totalSize          uint64
+	lastWrite          time.Time
+	maxSegmentSize     int
+	cache              map[string][][]byte
+	// this cache is a temporary placeholder to keep data while its being flushed
+	// and compacted. It's for cursors to combine the cache and this if a flush is occuring
+	flushCache        map[string][][]byte
+	cacheDirtySort    map[string]bool // will be true if the key needs to be sorted
+	cacheSizes        map[string]int
+	compactionRunning bool
+}
```

Comment:
I think `partition` should be exported so it can be tested. It's pretty isolated and it doesn't depend on the `Log`.